storage/pebble/operation/common.go

```diff
@@ -502,66 +327,93 @@ func traverse(prefix []byte, iteration iterationFunc) func(*badger.Txn) error {
 // removeByPrefix removes all the entities if the prefix of the key matches the given prefix.
 // if no key matches, this is a no-op
 // No errors are expected during normal operation.
-func removeByPrefix(prefix []byte) func(*badger.Txn) error {
-	return func(tx *badger.Txn) error {
-		opts := badger.DefaultIteratorOptions
-		opts.AllVersions = false
-		opts.PrefetchValues = false
-		it := tx.NewIterator(opts)
-		defer it.Close()
+func removeByPrefix(prefix []byte) func(pebble.Writer) error {
+	return func(tx pebble.Writer) error {
+		start, end := getStartEndKeys(prefix)
+		return tx.DeleteRange(start, end, nil)
+	}
+}
 
-		for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
-			key := it.Item().KeyCopy(nil)
-			err := tx.Delete(key)
-			if err != nil {
-				return irrecoverable.NewExceptionf("could not delete item with prefix: %w", err)
-			}
+// getStartEndKeys calculates the start and end keys for a given prefix.
+// note: DeleteRange takes [start, end) (inclusion on start, exclusive on end),
+func getStartEndKeys(prefix []byte) (start, end []byte) {
+	start = prefix
+
+	// End key is the prefix with the last byte incremented by 1
+	end = append([]byte{}, prefix...)
+	for i := len(end) - 1; i >= 0; i-- {
+		if end[i] < 0xff {
+			end[i]++
+			break
+		}
+		end[i] = 0
+		if i == 0 {
+			end = append(end, 0)
 		}
+	}
 
-		return nil
+	return start, end
+}
```

Comment:
## ⚠️ 

I am very skeptical that this is actually correct. My understanding is that if you call `removeByPrefix(0xFF)` for example, your computation of `getStartEndKeys` will do [something wired here](https://github.com/onflow/flow-go/blob/51969ecf92ea9775622890d6be9be223f6f7a593/storage/pebble/operation/common.go#L350-L352). So essentially, the values for the [`DeleteRange` call](https://github.com/onflow/flow-go/blob/51969ecf92ea9775622890d6be9be223f6f7a593/storage/pebble/operation/common.go#L333) are then:
```
start: [255]
end: [0 0]
```
Note that in this case, the `start` value is _bigger_ than the end value, and Pebble does _not delete anything_ I believe. I think your `DeleteRange` implementation is broken for _any_ prefix that is byte vector consisting solely of 1s. 

It is super important to rigorously test those boundary edge cases, but the tests don't cover boundary values 😞.

Reply:
I implemented a test to verify my understanding that this implementation is broken for this edge case, which confirms the current code is bugged:

```golang
// We test removing all keys with the same prefix utilizing 3-byte long keys:
//  - the first byte is the storage code (such as `codeServiceEvent`)
//  - second and this byte represent a reduced form of the blockID and EntityID respectively
//  - we want to delete all entries for a given block, i.e. as prefix we are providing the first 2 bytes
t.Run("prefix at boundary", func(t *testing.T) {
  storageCode := byte(0xFF)

  blockIDs := []int{255, 254, 253, 0, 1}
  blockId2Delete := byte(0xFF)

  unittest.RunWithPebbleDB(t, func(db *pebble.DB) {
    for _, blockID := range blockIDs {
      fmt.Println(blockID)
      for entityID := 255; entityID >= 0; entityID-- {
        value := uint64(blockID<<8 + entityID)
        key := []byte{storageCode, byte(blockID), byte(entityID), 0}
        writeValue(t, db, key, value)
      }
    }

    delPrefix := []byte{storageCode, blockId2Delete}
    err := removeByPrefix(delPrefix)(db)
    require.NoError(t, err)

    for _, blockID := range blockIDs {
      fmt.Println(blockID)
      for entityID := 255; entityID >= 0; entityID-- {
        key := []byte{storageCode, byte(blockID), byte(entityID), 0}
        expectedValue := uint64(blockID<<8 + entityID)

        var act Entity
        err = retrieve(key, &act)(db)

        if byte(blockID) == blockId2Delete {
          require.Error(t, err, "expecting error when retrieving entities for deleted block %d", blockID)
          assert.ErrorIs(t, convertNotFoundError(err), storage.ErrNotFound)
        } else {
          require.NoError(t, err)
          require.Equal(t, expectedValue, act.ID)
        }
      }
    }
  })
})
```

Reply:
Pebble has an [example in their documentation](https://pkg.go.dev/github.com/cockroachdb/pebble#example-Iterator-PrefixIteration) implementing this prefix logic that we can use, at least as a starting point to test these edge cases against. Given they make the database, there's a good chance that snippet is correct as-is.